### ISSUE TITLE
SSH Configuration issues

### ISSUE THREAD

**Issue by joebitcoinorg** [turnkeylinux/tracker#737]
I'm trying to get SSH reachable from the Internet.
Docker image  turnkey-joomla3-14.1-jessie-amd64

When people try to ssh to the machine this happens

> ssh root@67.52.2.70
Permission denied (publickey).

This hasn't changed even since modifying /etc/ssh/sshd_config  to enable:

PermitRootLogin yes
PasswordAuthentication yes
StrictModes yes

...by using this configuration line.

SSHD_OPTS="-f /etc/ssh/sshd_config -o StrictModes=yes -o PermitRootLogin=yes -o PasswordAuthentication=yes"

When I would inspect the process I saw no hint of where the active configuration was coming from... I dug deeper and found that  **/etc/default/ssh could be used to feed options that would show up in ps **auxw.****

So I put this into /etc/default/ssh:

SSHD_OPTS="-f /etc/ssh/sshd_config -o PermitRootLogin=yes -o PasswordAuthentication=yes"

Now ps auxw shows these options, however the same error message occurs.  Is it possible that sshd was compiled without these features?

Here's some error logs in /var/log/auth.log

/var/log/auth.log:Nov 23 17:42:31 a8cc04704fd1 login[20523]: pam_unix(login:auth): authentication failure; logname=root uid=0 euid=0 tty=/dev/pts/0 ruser= rhost=127.0.0.1 
/var/log/auth.log:Nov 23 17:42:34 a8cc04704fd1 login[20523]: FAILED LOGIN (4) on '/dev/pts/0' from '127.0.0.1' FOR 'UNKNOWN', Authentication failure
/var/log/auth.log:Nov 23 17:45:33 a8cc04704fd1 login[20556]: pam_unix(login:auth): authentication failure; logname=root uid=0 euid=0 tty=/dev/pts/0 ruser= rhost=127.0.0.1  user=root
/var/log/auth.log:Nov 23 17:45:37 a8cc04704fd1 login[20556]: FAILED LOGIN (1) on '/dev/pts/0' from '127.0.0.1' FOR 'root', Authentication failure
/var/log/auth.log:Nov 23 17:47:18 a8cc04704fd1 login[20618]: pam_unix(login:auth): authentication failure; logname=root uid=0 euid=0 tty=/dev/pts/0 ruser= rhost=127.0.0.1  user=root
/var/log/auth.log:Nov 23 17:47:22 a8cc04704fd1 login[20618]: FAILED LOGIN (1) on '/dev/pts/0' from '127.0.0.1' FOR 'root', Authentication failure
/var/log/auth.log:Nov 23 17:51:30 a8cc04704fd1 login[20773]: pam_unix(login:auth): authentication failure; logname=root uid=0 euid=0 tty=/dev/pts/0 ruser= rhost=127.0.0.1  user=root
/var/log/auth.log:Nov 23 17:51:34 a8cc04704fd1 login[20773]: FAILED LOGIN (1) on '/dev/pts/0' from '127.0.0.1' FOR 'root', Authentication failure
/var/log/auth.log:Nov 23 17:53:43 a8cc04704fd1 login[20899]: pam_unix(login:auth): authentication failure; logname=root uid=0 euid=0 tty=/dev/pts/0 ruser= rhost=127.0.0.1  user=root
/var/log/auth.log:Nov 23 17:53:45 a8cc04704fd1 login[20899]: FAILED LOGIN (1) on '/dev/pts/0' from '127.0.0.1' FOR 'root', Authentication failure
g

I have a volunteer webmaster who's asking for password style logins via ssh to use SCP. 

Thanks for looking into this.

**Comment by JedMeister** [comment]
That's really weird. When I tested the v14.1 builds (prior to release) I just used the same instructions as per the [original blog post announcement](https://www.turnkeylinux.org/blog/announcing-docker-builds) (obviously adjusting for v14.1 where necessary). I only tested Core thoroughly but password SSH login definitely worked (I rarely ever bother setting up keys for my tests). As all appliances inherit from Core, at least in theory, if SSH works on core it should work for all appliances.

The only thing that occurs to me is that if SSH is running on the host too, then did you configure port forwarding to an alternate port? The Docker containers fully expose ports to the host, but not to the network (unless you configure it). Perhaps you are inadvertently trying to connect to your host rather than the docker guest? I'm not sure on that though as your log is clearly showing authentication failures...

FWIW if you use a non-standard port (e.g. 2222) then you'll need to explicitly tell `ssh` to use that. E.g. 

````
ssh -p 2222 root@xx.xx.xx.xx
````

IIRC the equivalent switch for `scp` is `-P` (i.e. capital P instead of lower case p).

The only other thing I can think of is that for some reason your host (or something else on the network) is interfering with the traffic in some other way.

**Comment by joebitcoinorg** [comment]
In this case the docker instance is connected to the same bridge the 
Ethernet is.  So the docker instance gets a real ip address from our 
external network.  All the other services work from the Turnkey Build.

I'll check on the main host's ssh daemon to be sure that it isn't 
competing for the same port and ip.

Thanks for listening to my experience.

Regards,
Joe Baker

**Comment by qrntz** [comment]
As it is, I can't really tell what is wrong. The connection is going to the right place if there are new entries showing up in the logs of the corresponding sshd host, though.

@joebitcoinorg, could you please try to get some more verbose logs for both the server and the client? This should be a matter of adding
```
LogLevel DEBUG3
```

to `/etc/ssh/sshd_config` and using
```
ssh -vvv -o DebugLevel=DEBUG3 user@host
```
as your client-side command. Note that at this level of verbosity, sensitive information might be present in the logs and I would advise to look through the output before pasting them here. This [OpenSSH troubleshooting](https://en.wikibooks.org/wiki/OpenSSH/Logging_and_Troubleshooting) page might also help.

**Comment by joebitcoinorg** [comment]
Forgive me.
This issue is closed.  Operator error - IE. My mistake.

The ssh port docker was using was 1022 .  I was trying to use port 22 
which was controlled by a LXC container which contained the docker 
container.

**Comment by joebitcoinorg** [comment]
I had an Ah ha moment when I looked at this report of the docker ps command below.  Docker somehow chose to use port 1022 instead of 22  so I was trying the wrong port.  I tested it by turning off the ssh daemon and tried the client connection attempt and to my astonishment, there was still an ssh daemon responding - this was from not the host, but the LXC container from which I was running docker inside.

docker ps
CONTAINER ID        IMAGE                       COMMAND                CREATED             STATUS              PORTS                                                                                                  NAMES
a8cc04704fd1        turnkeylinux/joomla3-14.1   "/usr/sbin/start.sh"   2 weeks ago         Up 2 days           0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp, 0.0.0.0:12320-12322->12320-12322/tcp, **0.0.0.0:1022->22/tcp**   rkares-joomla3
